### DISCUSSION TITLE
[8.0 QA] react-docgen as TS default

### DISCUSSION THREAD

**Discussion by vanessayuenn** [storybookjs/storybook#25686] [Announcement]
#### Feature owner: @shilman 
Please comment in this discussion thread for any bug report and feedback for this feature as you test out Storybook 8. Kindly adhere to the following format:

> **Project**: Name of your project as entered in Notion
> **Version**: The version of storybook you tested with
> **Bug/feedback**: Details of the bugs/feedback you have

Feature owner should triage the comments here frequently and open issues as needed.

**Comment by yannbf** [comment]
**Project:** MealDrop
**Version:** 8.0.0-alpha.14
**Bug/feedback:** 

Types of the property `typescript` in `main.ts` are incompatible when using `react-docgen`.

Details and reproduction are provided in this issue:
https://github.com/storybookjs/storybook/issues/25736

**Comment by yannbf** [comment]
**Project:** MealDrop
**Version:** 8.0.0-alpha.15
**Bug/feedback:**

The automigration message has a few typos and seems confusing. At least to me, I was not sure what the migration wanted me to do. Seems more like a "do X or do Y" rather than a "yes/no" question to me.

> which dramatically faster but doesn't handle all TypeScript constructs.

which **is** dramatically

> though we recommend giving "react-docgen" for a much faster dev experience.

though we recommend giving "react-docgen" **a try** for a much faster dev experience.

**Repro:**

1. Add this to main.js:
```
typescript: {
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      propFilter: (prop) => {
        if (prop.parent) {
          return !prop.parent.fileName.includes('node_modules')
        }
        return true
      },
    }
  },
```
2. run `npx storybook@next automigrate`

**Comment by yannbf** [comment]
**Project:** MealDrop
**Version:** 8.0.0-alpha.14
**Bug/feedback:**

Not related to `react-docgen` but `react-docgen-typescript`. Seems like `react-docgen-typescript` does not have sensible defaults for its options. Is this intended?

For instance, without `reactDocgenTypescriptOptions`:
```ts
const config: StorybookConfig = {
  typescript: {
    reactDocgen: 'react-docgen-typescript',
  },
}
```

The types of all controls are inferred correctly
<img src="https://github.com/storybookjs/storybook/assets/1671563/39823f10-4293-465a-a66f-3ba3763c5629" width="300" />

With an empty object `reactDocgenTypescriptOptions` (or another setting, e.g. `shouldExtractLiteralValuesFromEnum: true`):
```ts
const config: StorybookConfig = {
  typescript: {
    reactDocgen: 'react-docgen-typescript',
    reactDocgenTypescriptOptions: { },
  },
}
```

Notice that everything becomes "object"
<img src="https://github.com/storybookjs/storybook/assets/1671563/e9fdaccf-b493-487e-821d-5db1b5657a21" width="300" />

I'd assume that users don't have to know which are the default values, so that if they set a single option such as `shouldExtractLiteralValuesFromEnum`, they would still get the previous behavior + the enum one

**Repro:**
It's hard because it will only happen once.
1. Go to https://github.com/yannbf/mealdrop/tree/storybook/8.0.0
2. Use the settings described above

**Reply by shilman** [reply]
if you want to merge a user-defined preset value with an existing preset you should define a preset function and merge it yourself:

```js
export const typescript = (existing) => ({
  ...existing,
  reactDocgenTypescriptOptions: {
    ...existing?.reactDocgenTypescriptOptions,
    /* some options here */
  },
});
```

It wouldn't be hard to change this so that the default options are empty and then the merge happens at the place where the options are consumed. However, this would be non-standard and also suck if we wanted to consume the data in various places.

@ndelangen @yannbf do you have any thoughts on this? It seems like basically a complaint about how presets work.

**Reply by ndelangen** [reply]
@shilman I'd say no to changing this. How WOULD you remove a default property then?

**Reply by yannbf** [reply]
I think it makes sense to merge the options. However, [the docs](https://storybook.js.org/docs/api/main-config-typescript) never mention anything regarding merging options, so I would assume users would always fall into this issue. It might seem obvious to the Storybook team but I think users might not really understand that this is the way to go. cc @kylegach WDYT?

**Comment by ndelangen** [comment]
Here's the default: 

https://github.com/storybookjs/storybook/blob/5e3532d91a3b3723e680da477bddbdcb6279991a/code/lib/core-server/src/presets/common-preset.ts#L140-L146

but we don't deeply merge when presets values are objects.
